internal/command/arguments/add.go

```diff
@@ -0,0 +1,128 @@
+package arguments
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+)
+
+// Add represents the command-line arguments for the Add command.
+type Add struct {
+	// Addr specifies which resource to generate configuration for.
+	Addr addrs.AbsResourceInstance
+
+	// FromResourceAddr specifies the address of an existing resource in state
+	// which should be used to populate the template.
+	FromResourceAddr *addrs.AbsResourceInstance
+
+	// OutPath contains an optional path to store the generated configuration.
+	OutPath string
+
+	// Optional specifies whether or not to include optional attributes in the
+	// generated configuration. Defaults to false.
+	Optional bool
+
+	// Provider specifies the provider for the target.
+	Provider addrs.Provider
```

Comment:
I'm kinda curious about the use-cases for this argument. I'm fine with including it if we can imagine situations where it would be important, but I'm asking because I believe we only had this argument on some other commands like `terraform import` because historically they didn't refer to configuration at all, and only later did we make them consult the configuration to see which provider to use; given that we're not constrained by backward compatibility on this new command, I'd lean towards leaving this out if we can, to have fewer usage variations to support.

Reply:
The use-case is the (unlikely!) scenario when you have multiple providers with the same type in a configuration. If I had a configuration with both `hashicorp/random` and `mildwonkey/random`, I'd need to tell terraform which provider I meant if it weren't the default (either an implied default, or the provider with a local name of "random").  

This is important because (I know you know this, but writing out for completeness) those providers could have entirely different schemas and resource types.

Reply:
now that I say this, the `google-beta` provider is another example of a more-plausible real world use case!

Reply:
Thanks! In retrospect I think I had my wires crossed when I originally wrote this comment, because I was still thinking about `terraform import` (adding things to the state) rather than `terraform add` (adding things to the config). But with fresh eyes this makes perfect sense.

This does lead me to a different question, though: in order to later fold the import functionality into this, we'll need to know a specific provider _configuration_ to use, rather than just a provider. We'd need it both to know which provider configuration to run the import operation with, and also to know what to write into the `provider` argument in the generated configuration to make sure it ends up associated with the same provider configuration that imported it.

Given that, I wonder if we should pre-emptively set up this argument as an `addrs.AbsProviderConfig` instead of just an `addrs.Provider`, so that we'll be ready for the later extension. The downside is that the syntax for those is a bit clunky to write on the command line: `-provider='module.bar.module.baz.provider["registry.terraform.io/hashicorp/aws"].foo'` ... but hopefully this option is rarely used enough that it being a bit clunky is okay in return for it being explicit.

Perhaps even in the short term we could use the more elaborate form of this to fill in the `provider` argument in the generated resource block, so that for my example above it would write `provider = aws.foo`. (In the `-from-state` case we can infer a full provider configuration address from the state, so I think for now it'd only be for the mode where it's only generating a placeholder skeleton.)

Reply:
🤔 That's is a _very_ good point, thank you.

`add` does already fill in the `provider` argument when the given provider doesn't match the implied provider, but - as you have just pointed out - it would be missing any alias. I'll make that change!

Reply:
I've refactored with this change and updated the tests.
One notable change is I've now made `-provider` and `-from-state` conflicting flags; if `-from-state` is selected we will get the provider directly from state (looking up the provider local name from configuration, as before, but now we know we're using the correct provider FQN).